core/tests/mod.rs

```diff
@@ -0,0 +1,646 @@
+                        .apply_wsv(WorldStateView::clone)
+                        .block_hashes_after_hash(None)[0]
+                );
+                return;
+            }
+        }
+        sleep_ms(100);
+    }
+    panic!("Sumeragi did not submit genesis.");
+}
+
+#[test]
+fn sumeragi_all_peers_commit_genesis() {
+    init_telemetry();
+
+    /* Can't have 1 peer, then block is not sent. Can't have < 4 peers because non genesis peers get upset. */
+    let (keypairs, peer_ids, configs, _queues, sumeragis) =
+        set_up_network_with_genesis(4, false, Arc::new(move |_packet, _peer_id, _sumeragis| {}));
+    let sumeragis = sumeragis.lock().unwrap();
+    let sumeragi = sumeragis.get(&peer_ids[0]).unwrap();
+    let other_key = keypairs[1].clone();
+    let other_config = configs[1].clone();
+
+    let genesis_block: VersionedCommittedBlock = {
+        let kura = Kura::blank_kura_for_testing();
+        let wsv = WorldStateView::new(genesis_world(peer_ids.clone()), Arc::clone(&kura));
+
+        assert_eq!(wsv.height(), 0);
+        assert_eq!(wsv.latest_block_hash(), None);
+
+        let genesis_network = GenesisNetwork::test(other_config, true).unwrap();
+        let transactions = genesis_network.transactions;
+        assert!(
+            !transactions.is_empty(),
+            "Genesis transaction set contains no valid transactions"
+        );
+
+        let block = BlockBuilder {
+            transactions,
+            event_recommendations: Vec::new(),
+            view_change_index: 0,
+            committed_with_topology: network_topology::Topology::new(peer_ids.clone()),
+            key_pair: other_key,
+            wsv,
+        }
+        .build();
+
+        assert!(
+            block.rejected_transactions.is_empty(),
+            "Genesis transaction set contains invalid transactions"
+        );
+
+        block.commit_unchecked().into()
+    };
+
+    sumeragi.incoming_message(MessagePacket::new(
+        ProofChain::default(),
+        Message::BlockSyncUpdate(genesis_block.clone().into()),
+    ));
+
+    for _ in 0..300 {
+        if sumeragi
+            .apply_wsv(WorldStateView::clone)
+            .block_hashes_after_hash(None)
+            .len()
+            == 1
+        {
+            assert_eq!(
+                genesis_block.hash(),
+                sumeragi
+                    .apply_wsv(WorldStateView::clone)
+                    .block_hashes_after_hash(None)[0]
+            );
+            return;
+        }
+        sleep_ms(100);
+    }
+    panic!("Sumeragi did not commit genesis correctly.");
+}
+
+#[test]
+fn sumeragi_no_faults_commit_blocks() {
+    init_telemetry();
+
+    let (_keypairs, peer_ids, _configs, queues, sumeragis) = set_up_network(
+        7,
+        Arc::new(move |packet, peer_id, sumeragis| {
+            let sumeragis = sumeragis.lock().unwrap();
+            sumeragis.get(peer_id).unwrap().incoming_message(packet);
+        }),
+    );
+
+    for block_height in 1..10 {
```

Comment:
Can you add some comments describing what's going on inside such a big blocks of code? I'd prefer splitting into functions, but I understand that it might be hard sometimes

Reply:
On that note, most tests need extra comments. 

You have to describe a strict set of preconditions with an example, a strict set of operations, and a strict set of postconditions. With something as complex as this, you might need to also elaborate on specific pre-conditions.

Reply:
What's concerning here is that Rust cannot differentiate `test` function code from the code that it's covering. You're pushing the coverage up, but you're not covering the code that's in use. Given such a complex test, it looks to me that you're effectively mocking what you should be testing instead. 

Consider if we can remove this in favour of #3490 ?